File: mayi/src/main/java/com/thanosfisherman/mayi/PermissionRationaleToken.java
package com.thanosfisherman.mayi;

import android.os.Build;
import android.support.annotation.RequiresApi;

import java.lang.ref.WeakReference;

final class PermissionRationaleToken implements PermissionToken
{
    private final WeakReference<MayiFragment> mayiFragment;
    private boolean isTokenResolved = false;

    PermissionRationaleToken(MayiFragment mayiFragment)
    {
        this.mayiFragment = new WeakReference<>(mayiFragment);
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    @Override
    public void continuePermissionRequest()
    {
        if (!isTokenResolved)
        {
            mayiFragment.get().onContinuePermissionRequest();
            isTokenResolved = true;
        }
    }

    @Override
    public void skipPermissionRequest()
    {
        if (!isTokenResolved)
        {
            mayiFragment.get().onSkipPermissionRequest();
            isTokenResolved = true;
        }
    }
}

File: mayi/src/main/java/com/thanosfisherman/mayi/PermissionRationaleToken.kt
package com.thanosfisherman.mayi

import android.os.Build
import androidx.annotation.RequiresApi

internal class PermissionRationaleToken(private val permissionToken: PermissionToken) : PermissionToken {

    private var isTokenResolved = false

    @RequiresApi(api = Build.VERSION_CODES.M)
    override fun continuePermissionRequest() {
        if (!isTokenResolved) {
            permissionToken.let {
                it.continuePermissionRequest()
                isTokenResolved = true
            }
        }
    }

    override fun skipPermissionRequest() {
        if (!isTokenResolved) {
            permissionToken.let {
                it.skipPermissionRequest()
                isTokenResolved = true
            }
        }
    }
}
